aptos-move/aptos-gas-algebra/src/abstract_algebra.rs

```diff
@@ -0,0 +1,273 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use either::Either;
+use move_core_types::gas_algebra::{GasQuantity, UnitDiv};
+use std::{
+    marker::PhantomData,
+    ops::{Add, Mul},
+};
+
+/***************************************************************************************************
+ * Gas Expression & Visitor
+ *
+ **************************************************************************************************/
+/// Trait representing an abstract expression that can be used to calculate some gas amount.
+///
+/// It carries a type parameter `E`, indicating an environment in which the expression can be
+/// evaluated/materialized.
+pub trait GasExpression<E> {
+    type Unit;
+
+    /// Evaluates the expression within the given environment to a concrete number.
+    fn evaluate(&self, feature_version: u64, env: &E) -> GasQuantity<Self::Unit>;
+
+    /// Traverse the expression in post-order using the given visitor.
+    /// See [`GasExpressionVisitor`] for details.
+    fn visit(&self, visitor: &mut impl GasExpressionVisitor);
+
+    /// Performs a division on the unit of the expression.
+    ///
+    /// This is sometimes required if you want to multiply an amount by a certain count.
+    fn per<U>(self) -> GasPerUnit<Self, U>
+    where
+        Self: Sized,
+    {
+        GasPerUnit {
+            inner: self,
+            phantom: PhantomData,
+        }
+    }
+}
+
+/// An interface for performing post-order traversal of the tree structure of a gas expression.
+///
+/// Alternatively, one could think that the callbacks are invoked following the Reverse Polish
+/// notation of the expression.
+///
+/// Here are a few examples:
+/// - `1 + 2`
+///   - `quantity(1)`
+///   - `quantity(2)`
+///   - `add()`
+/// - `A + B * 50`
+///   - `gas_param<A>()`
+///   - `gas_param<B>()`
+///   - `quantity(50)`
+///   - `mul()`
+///   - `add()`
+pub trait GasExpressionVisitor {
```

Comment:
is this actually used? I don't see any reference on this

Reply:
It's not used in this PR, but @wlawt's gas calibrator will use this interface do record abstract gas usage.